src/base_classes/active_space_solver.cc

```diff
@@ -26,72 +26,423 @@
  * @END LICENSE
  */
 
+#include <numeric>
+#include <iomanip>
+
+#include "psi4/psi4-dec.h"
+#include "psi4/libpsi4util/PsiOutStream.h"
+#include "psi4/libmints/molecule.h"
+#include "psi4/libmints/pointgrp.h"
 #include "psi4/libmints/wavefunction.h"
+#include "psi4/libpsi4util/process.h"
 
-#include "base_classes/mo_space_info.h"
 #include "base_classes/forte_options.h"
-#include "fci/fci_solver.h"
-#include "casscf/casscf.h"
-#include "sci/aci.h"
-#include "sci/asci.h"
-#include "sci/fci_mo.h"
+#include "base_classes/reference.h"
+#include "base_classes/mo_space_info.h"
+#include "helpers/helpers.h"
+#include "integrals/active_space_integrals.h"
+#include "active_space_method.h"
 
-#include "base_classes/active_space_solver.h"
+#include "active_space_solver.h"
 
 namespace forte {
 
-ActiveSpaceSolver::ActiveSpaceSolver(StateInfo state, std::shared_ptr<MOSpaceInfo> mo_space_info,
-                                     std::shared_ptr<ActiveSpaceIntegrals> as_ints)
-    : states_weights_({{state, 1.0}}), mo_space_info_(mo_space_info), as_ints_(as_ints) {
-    active_mo_ = as_ints_->active_mo();
-    core_mo_ = as_ints_->restricted_docc_mo();
+ActiveSpaceSolver::ActiveSpaceSolver(
+    const std::string& method,
+    std::vector<std::pair<StateInfo, std::vector<double>>>& state_weights_list,
+    std::shared_ptr<SCFInfo> scf_info, std::shared_ptr<MOSpaceInfo> mo_space_info,
+    std::shared_ptr<ActiveSpaceIntegrals> as_ints, std::shared_ptr<ForteOptions> options)
+    : method_(method), state_weights_list_(state_weights_list), scf_info_(scf_info),
+      mo_space_info_(mo_space_info), as_ints_(as_ints), options_(options) {
+    print_options();
 }
 
-ActiveSpaceSolver::ActiveSpaceSolver(
-    const std::vector<std::pair<StateInfo, double>>& states_weights,
-    std::shared_ptr<MOSpaceInfo> mo_space_info, std::shared_ptr<ActiveSpaceIntegrals> as_ints)
-    : states_weights_(states_weights), mo_space_info_(mo_space_info), as_ints_(as_ints) {}
+double ActiveSpaceSolver::compute_energy() {
+    double energy = 0.0;
+    for (const auto& [state, weights] : state_weights_list_) {
+        // compute the energy of state and save it
+        size_t nroot = weights.size();
+        std::shared_ptr<ActiveSpaceMethod> method = make_active_space_method2(
+            method_, state, nroot, scf_info_, mo_space_info_, as_ints_, options_);
+
+        method->set_options(options_);
+        if( set_rdm_ ){
+            method->set_max_rdm_level(max_rdm_level_);
+        }
+        method->compute_energy();
+        auto energies = method->energies();
+        for (size_t r = 0; r < nroot; r++) {
+            energy += energies[r] * weights[r];
+        }
+        method_vec_.push_back(method);
+    }
+    return energy;
+}
+
+Reference ActiveSpaceSolver::get_reference(int root) {
```

Comment:
I suppose we need this for back compatibility but we can at least start to do a couple of things to make the purpose of this function clearer:

- wrap the state-averaged procedure in its own function that is called on line 91.
- put the code that transforms the RDM <-> cumulant in a separate function in the reference.cc file. This should go in a free function and not in the class.
- the input of this functions should be generalized to get a pair (state symmetry, root). But right now it is more important to pass the root variable down to the other methods.

Reply:
I have some un-pushed commits that do much of this work. Currently, the SA references are built properly using the ActiveSpaceSolver interface (even for FCI_MO!), but the final computation of the energy/oscillator strengths does not work since each DSRG method handles that separately at the moment. The issue is that that part of the code needs knowledge of the determinants for each state. A quick fix is to implement a method->determinants() function, though generally I don't think we want this.